### PR TITLE
rasterizer_cache: Remove runtime allocation caching

### PR DESCRIPTION
**Pull request by GPUCode**
This is required for vulkan to be back in canary. To explain in more detail, since the start Citra has always kept OpenGL textures cached in some form or another, without freeing them. Before #5710 surfaces themselves would be kept forever, while after that PR textures would be kept in a separate allocation cache. This was actually required because not doing so would cripple performance in all games.

That's because there are some surfaces that get destroyed and recreated each frame. Some examples:
- If the game inserts a fill surface to a clear a framebuffer, the fill surface invalidates the region and takes ownership of it from the framebuffer, which destroys it. Later the game requests the framebuffer in GetFramebuffer surfaces, so the cache recreates it and validates it from the fill surface.
- During format reinterpretation (for example D24S8 -> RGBA8) an RGBA8 surface is created when GetTextureSurface is called and is validated from the D24S8 one. Next draw however the depth buffer is invalidated causing the RGBA8 to get removed.
- FMVs constantly stream data to a memory location, causing CPU side invalidation and destruction of the owner surface.

Initially I attempted a simple solution of marking some surfaces with a "dont destroy" flag which would prevent the cache from outright freeing it. Said solution though was unable to handle FMVs so it had to scrapped. The alternative is implemented here, a simple garbage collector that destroys sentenced surfaces. When a surface becomes fully invalidated and is not a fill surface (these are destroyed always because AccelerateFill will always insert a new one currently. Should this changed in the future?) it is marked as sentenced instead of being removed. The sentence flag can only be removed if a FindMatch references that surface.

Every 8 ticks the garbage collectors checks which senteced surfaces have surpassed the removal threshold and removes them if so. The threshold has been specially chosen to be greater than the swapchain image [count](https://github.com/citra-emu/citra/blob/700c00f021dbe7674a42971d7ca3c3916d3b6b41/src/video_core/renderer_opengl/renderer_opengl.cpp#L41) in the OpenGL mailbox, as any value lower than that would cause presentation artifacts. I'm not sure if the OpenGL spec guarantees synchronization between shared contexts, but better be safe here. The garbage collector only works on sentenced surfaces. Unless the game explicitely invalidated a region we should not remove any surfaces on that region.

From my testing this hasn't affected performance at all. One small downside is that new textures are created on scene transitions, though it doesn't appear to be noticeable. In contrast allocation caching would quickly end up with dozen of unused allocations just from normal playthroughs. The best case I saw was 50 unused allocations after a few minutes, while some games like pokemon can reach 200 unused allocations just from switching areas.

Another point to consider is that our recycle rate is inflated currently because we decode all texture surfaces to RGBA8, which gives more opportunities for reuse. If we were to switch to using swizzles of native formats however, which is something we should do to reduce memory bandwidth on mobile, allocation caching would become even more specialized and not so useul.

Moreover, I mentioned Vulkan at the start. The OpenGL driver can handle having too much allocations by moving the less used ones to system RAM. Vulkan however does no such thing, which means that after a few minutes most games will crash because VMA exceeded the memory budget or there was an out of memory pipeline error.

Thorough testing is appreciated as the cache may still have behaviourst that rely on the same allocation being recycled when it shouldn't

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6705)

### CODE REVIEW

**Comment by GPUCode** [comment]
Seems there are some issues with CPU writes, marking as draft

**Comment by FernandoS27** [comment]
Why not use an LRU cache with a frame id instead? Every time the texture is reused you remove it from the lru. Every swap, you check the LRU, the textures 6 frames behind get removed. In yuzu we have a class for LRU caching.

**Comment by GPUCode** [comment]
The post is slightly outdated for the garbage collection method but that's basically what it does now, just with a single list instead of using DelayedDestructionRing.

**Review by Steveice10 — APPROVED** [review]
Assuming it works well in practice, the code looks good to me.

**Comment by GPUCode** [comment]
The only thing that's preventing it from going to canary is reworking framebuffer caching to not rely on OpenGL handles. Will tag it when I come up with a clean solution to that.

Also while at it, I would like to document some of the newer changes. The resolution scaling adjustment code was removed from GetSurface as surfaces now can automatically be rescaled when needed. The first check ensured that a subrect of a resolution scaled texture would stay resolution scaled. This is now handled in CopySurfaces, where we check if the dest scale is smaller than the small scale (down-sample) and up scale the dest surface in that case to prevent scaled data loss.

The reinterpretation check has been similarly moved to ValidateByReinterpretation and is now also more generic and applies to all possible reinterpretations, not only Pokemon.

I've also removed the Expand flag as the above removals limited its usage and I could not find a game where it provides a benefit. From what I could see, the name is a bit misleading. It doesn't take a surface and make it larger. Rather it merges two surfaces that overlap and have the same stride. In theory it can benefit some games by skipping a copy, but real world shows the opposite: removing the flag removes 4+ texture copies in Monster Hunter Stories increasing performance. If any game comes up that slows down due to the removal of the flag, it can be readded.

Final change is using texture copy instead of blit in CopySurfaces when resolution scale matches. This is the most common case and should benefit mali GPUs as texture blits are very expensive there

File: src/video_core/CMakeLists.txt [hunk 1]
```diff
@@ -132,6 +133,7 @@ add_library(video_core STATIC
     vertex_loader.h
     video_core.cpp
     video_core.h
+    renderer_opengl/gl_texture_mailbox.h renderer_opengl/gl_texture_mailbox.cpp
```
Comment (Steveice10) [inline src/video_core/CMakeLists.txt:136]: Duplicate?

**Review by Steveice10 — APPROVED** [review]
Re-approving with latest revisions. Also tested a few games in canary and didn't find any new issues.

**Pull request merged**